torchvision/prototype/datasets/utils/_dataset.py

```diff
@@ -0,0 +1,218 @@
+
+    def __delattr__(self, item: Any) -> NoReturn:
+        raise RuntimeError(f"'{type(self).__name__}' object is immutable")
+
+    def __hash__(self) -> int:
+        return self.__dict__["__final_hash__"]
+
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, DatasetConfig):
+            return NotImplemented
+
+        return hash(self) == hash(other)
+
+    def __repr__(self) -> str:
+        return make_repr(type(self).__name__, self.items())
+
+
+class DatasetInfo:
+    def __init__(
+        self,
+        name: str,
+        *,
+        categories: Union[int, Sequence[str], str, pathlib.Path],
+        citation: Optional[str] = None,
+        homepage: Optional[str] = None,
+        license: Optional[str] = None,
+        valid_options: Optional[Dict[str, Sequence]] = None,
+    ) -> None:
+        self.name = name.lower()
+
+        if isinstance(categories, int):
+            categories = [str(label) for label in range(categories)]
+        elif isinstance(categories, (str, pathlib.Path)):
+            with open(pathlib.Path(categories).expanduser().resolve(), "r") as fh:
+                categories = fh.readlines()
+        self.categories = categories
+
+        self.citation = citation
+        self.homepage = homepage
+        self.license = license
+
+        valid_split: Dict[str, Sequence] = dict(split=["train"])
+        if valid_options is None:
+            valid_options = valid_split
+        elif "split" not in valid_options:
+            valid_options.update(valid_split)
+        elif "train" not in valid_options["split"]:
+            raise ValueError(
+                f"'train' has to be a valid argument for option 'split', "
+                f"but found only {sequence_to_str(valid_options['split'], separate_last='and ')}."
+            )
+        self._valid_options: Dict[str, Sequence] = valid_options
+
+    @property
+    def default_config(self) -> DatasetConfig:
+        return DatasetConfig(
+            {name: valid_args[0] for name, valid_args in self._valid_options.items()}
+        )
+
+    def make_config(self, **options: Any) -> DatasetConfig:
+        for name, arg in options.items():
+            if name not in self._valid_options:
+                raise ValueError(
+                    add_suggestion(
+                        f"Unknown option '{name}' of dataset {self.name}.",
+                        word=name,
+                        possibilities=sorted(self._valid_options.keys()),
+                    )
+                )
+
+            valid_args = self._valid_options[name]
+
+            if arg not in valid_args:
+                raise ValueError(
+                    add_suggestion(
+                        f"Invalid argument '{arg}' for option '{name}' of dataset {self.name}.",
+                        word=arg,
+                        possibilities=valid_args,
+                    )
+                )
```

Comment:
We can move this to a dedicated function, so we have the ability to override the default option checking.